transport/tlsfrag/buffer.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlsfrag
+
+import (
+	"errors"
+	"fmt"
+	"io"
+)
+
+var (
+	// errTLSClientHelloFullyReceived is returned when a full TLS Client Hello has been received and no
+	// more data can be pushed to the buffer.
+	errTLSClientHelloFullyReceived = errors.New("already received a complete TLS Client Hello packet")
+
+	// errInvalidTLSClientHello is the error used when the data received is not a valid TLS Client Hello.
+	// Please use [errors.Is] to compare the returned err object with this instance.
+	errInvalidTLSClientHello = errors.New("not a valid TLS Client Hello packet")
+)
+
+// clientHelloBuffer is a byte buffer used to receive and buffer a TLS Client Hello packet.
+type clientHelloBuffer struct {
+	data   []byte // the buffer that hosts both header and content, len: 5 -> 5+len(content)
+	len    int    // the actual bytes that have been read into data
+	valid  bool   // indicate whether the content in data is a valid TLS Client Hello record
+	toRead int    // the number of bytes to read next, e.g. 5 -> len(content)
+}
+
+var _ io.Writer = (*clientHelloBuffer)(nil)
+var _ io.ReaderFrom = (*clientHelloBuffer)(nil)
+
+// newClientHelloBuffer creates and initializes a new buffer to receive a TLS Client Hello packet.
+func newClientHelloBuffer() *clientHelloBuffer {
+	// Allocate the 5 bytes header first, and then reallocate it to contain the entire packet later
+	return &clientHelloBuffer{
+		data:   make([]byte, recordHeaderLen),
```

Comment:
I tried that before, and it will cause additional buffer allocations in `ReadFrom`.

Think of the scenario reading from `r io.Reader`, you need to call `r.Read(p)`, and `p` cannot be empty. If you use `r.Read(data[:5])`, the `data` will be expanded to 5 bytes immediately, what if `r` doesn't have that many bytes available (like only 3 bytes is in `r`), when `len(b.data) == 5` but actually there are only 3 valid bytes read.